clang/lib/Sema/BoundsWideningAnalysis.cpp

```diff
@@ -14,4 +14,396 @@
+                                               BinaryOperatorKind::BO_Add);
+}
+
+Expr *BoundsWideningAnalysis::GetTerminatorCondition(const Expr *E) const {
+  if (!E)
+    return nullptr;
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E->IgnoreParens()))
+    return GetTerminatorCondition(BO->getRHS());
+
+  // According to C11 standard section 6.5.13, the logical AND Operator
+  // shall yield 1 if both of its operands compare unequal to 0;
+  // otherwise, it yields 0. The result has type int.
+  // If we have if (*p && *(p + 1)) where p is _Nt_array_ptr<char> then
+  // it is casted to integer type and an IntegralCast is generated. Here
+  // we strip off the IntegralCast.
+  if (auto *CE = dyn_cast<CastExpr>(E))
+    if (CE->getCastKind() == CastKind::CK_IntegralCast)
+      return const_cast<Expr *>(CE->getSubExpr());
+  return const_cast<Expr *>(E);
+}
+
+Expr *BoundsWideningAnalysis::GetTerminatorCondition(const CFGBlock *B) const {
+  if (!B)
+    return nullptr;
+
+  if (const Stmt *S = B->getTerminatorStmt()) {
+    if (const auto *BO = dyn_cast<BinaryOperator>(S))
+      return GetTerminatorCondition(BO->getLHS());
+
+    if (const auto *IfS = dyn_cast<IfStmt>(S))
+      return GetTerminatorCondition(IfS->getCond());
+
+    if (const auto *WhileS = dyn_cast<WhileStmt>(S))
+      return const_cast<Expr *>(WhileS->getCond());
+
+    if (const auto *ForS = dyn_cast<ForStmt>(S))
+      return const_cast<Expr *>(ForS->getCond());
+
+    if (const auto *SwitchS = dyn_cast<SwitchStmt>(S))
+      return const_cast<Expr *>(SwitchS->getCond());
+  }
+  return nullptr;
+}
+
+Expr *BoundsWideningAnalysis::GetDerefExpr(Expr *E) const {
+  // A dereference expression can contain an array subscript or a pointer
+  // dereference.
+  if (!E)
+    return nullptr;
+
+  if (auto *CE = dyn_cast<CastExpr>(E))
+    if (CE->getCastKind() == CastKind::CK_LValueToRValue)
+      E = CE->getSubExpr();
+
+  E = IgnoreCasts(E);
+
+  // If a dereference expression is of the form "*(p + i)".
+  if (auto *UO = dyn_cast<UnaryOperator>(E)) {
+    if (UO->getOpcode() == UO_Deref)
+      return IgnoreCasts(UO->getSubExpr());
+
+  // Else if a dereference expression is an array access. An array access can
+  // be written A[i] or i[A] (both are equivalent).  getBase() and getIdx()
+  // always present the normalized view: A[i]. In this case getBase() returns
+  // "A" and getIdx() returns "i". We normalize A[i] to "A + i" and return that
+  // as the dereference expression.
+  } else if (auto *AE = dyn_cast<ArraySubscriptExpr>(E)) {
+    return ExprCreatorUtil::CreateBinaryOperator(SemaRef, AE->getBase(),
+                                                 AE->getIdx(),
+                                                 BinaryOperatorKind::BO_Add);
+  }
+  return nullptr;
+}
+
+void BoundsWideningAnalysis::GetVarsInExpr(Expr *E,
+                                           VarSetTy &VarsInExpr) const {
+  // Get the VarDecls of all variables occurring in an expression.
```

Comment:
It seems like `GetVarsInExpr` returns a subset of all variables occurring in an expression. For example, in the expression `s->f + myFunction(x)`, `GetVarsInExpr` will return an empty set, not `{ s, x }`. Could this function include a comment stating that it only returns some of the variables that appear in `E`?